### ISSUE TITLE
Recovering from transaction log wipeout

### ISSUE THREAD

**Issue by tricepat** [neo4j/neo4j#12115]
I accidentally removed all neostore.transaction files. My neo4j instance is now in a restarting loop, with the error: No check point found in any log file from version 14 to 14. 

I am reading now that at least one transaction file is required for recovery (I assume the shutdown was not clean). Given I do not have the checkpoints anymore, how do I resolve this?

**Neo4j Version:** 3.4.1
**Operating System:** Ubuntu 18.04.1
**API:** Python driver

### Steps to reproduce
1. Remove all neostore.transaction.* files 
2. Restart neo4j "systemctl restart neo4j"

### Expected behavior
Neo4j initializes and starts.

### Actual behavior
Neo4j initializes but does not start

### Logs
org.neo4j.server.ServerStartupException: Starting Neo4j failed: Component 'org.neo4j.server.database.LifecycleManagingDatabase@7df60067' was successfully initialized, but failed to start. Please see the attached cause exception "No check point found in any log file from version 14 to 14".
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.exception.ServerStartupErrors.translateToServerStartupError(ServerStartupErrors.java:68)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:220)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:111)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.ServerBootstrapper.start(ServerBootstrapper.java:79)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at com.neo4j.server.enterprise.CommercialEntryPoint.main(CommercialEntryPoint.java:22)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.server.database.LifecycleManagingDatabase@7df60067' was successfully initialized, but failed to start. Please see the attached cause exception "No check point found in any log file from version 14 to 14".
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:466)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.AbstractNeoServer.start(AbstractNeoServer.java:212)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         ... 3 more
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: Caused by: java.lang.RuntimeException: Error starting org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory, /var/lib/neo4j/data/databases/graph.db
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:212)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.enterprise.EnterpriseGraphDatabase.<init>(EnterpriseGraphDatabase.java:39)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.enterprise.OpenEnterpriseNeoServer.lambda$static$1(OpenEnterpriseNeoServer.java:78)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.server.database.LifecycleManagingDatabase.start(LifecycleManagingDatabase.java:88)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:445)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         ... 5 more
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.NeoStoreDataSource@383f1975' was successfully initialized, but failed to start. Please see the attached cause exception "No check point found in any log file from version 14 to 14".
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:466)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:100)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:445)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:107)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.impl.factory.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:208)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         ... 9 more
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: Caused by: java.lang.RuntimeException: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.recovery.Recovery@58cec85b' failed to initialize. Please see the attached cause exception "No check point found in any log file from version 14 to 14".
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:531)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:445)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         ... 14 more
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: Caused by: org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.recovery.Recovery@58cec85b' failed to initialize. Please see the attached cause exception "No check point found in any log file from version 14 to 14".
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:427)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport.init(LifeSupport.java:62)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:98)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:515)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         ... 15 more
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: Caused by: org.neo4j.kernel.impl.store.UnderlyingStorageException: No check point found in any log file from version 14 to 14
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.recovery.RecoveryStartInformationProvider.get(RecoveryStartInformationProvider.java:108)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.recovery.DefaultRecoveryService.getRecoveryStartInformation(DefaultRecoveryService.java:61)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.recovery.Recovery.init(Recovery.java:67)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:406)
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]:         ... 18 more
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21019]: 2018-12-21 17:58:30.207+0000 INFO  Neo4j Server shutdown initiated by request
Dec 21 17:58:30 ip-172-31-36-134 systemd[1]: neo4j.service: Main process exited, code=exited, status=1/FAILURE
Dec 21 17:58:30 ip-172-31-36-134 systemd[1]: neo4j.service: Failed with result 'exit-code'.
Dec 21 17:58:30 ip-172-31-36-134 systemd[1]: neo4j.service: Service hold-off time over, scheduling restart.
Dec 21 17:58:30 ip-172-31-36-134 systemd[1]: neo4j.service: Scheduled restart job, restart counter is at 840.
Dec 21 17:58:30 ip-172-31-36-134 systemd[1]: Stopped Neo4j Graph Database.
Dec 21 17:58:30 ip-172-31-36-134 systemd[1]: Started Neo4j Graph Database.
Dec 21 17:58:30 ip-172-31-36-134 pre-neo4j.sh[21150]: pre-neo4j.sh: Fetching AWS instance metadata
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: Unable to locate credentials. You can configure credentials by running "aws configure".
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: Unable to locate credentials. You can configure credentials by running "aws configure".
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_https_enabled true
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_https_listen_address 0.0.0.0:7473
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_http_enabled true
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_http_listen_address 0.0.0.0:7474
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_bolt_enabled true
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_bolt_listen_address 0.0.0.0:7687
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connector_bolt_tls_level REQUIRED
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_backup_enabled true
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_backup_address localhost:6362
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: causal_clustering_discovery_type LIST
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: causal_clustering_initial_discovery_members localhost:5000
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: causal_clustering_minimum_core_cluster_size_at_formation 3
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: causal_clustering_minimum_core_cluster_size_at_runtime 3
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_connectors_default_listen_address 0.0.0.0
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_mode SINGLE
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: causal_clustering_discovery_listen_address 0.0.0.0:5000
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_logs_http_enabled false
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_logs_gc_enabled false
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_logs_security_level INFO
Dec 21 17:58:31 ip-172-31-36-134 pre-neo4j.sh[21150]: dbms_security_allow_csv_import_from_file_urls true
...

**Comment by MishaDemianenko** [comment]
Hi @tricepat, thx for Your report.

I have tried to reproduce reported behaviour over multiple version (different 3.4 and 3.5 versions) and so far I do not see a way that will prevent database to start in the case when transaction logs are not there.

Also, please note that in the provided stacktrace You can see the message "No check point found in any log file from version 14 to 14". That means that tx log file with index 14 was present during the database start.
Since that file did not contain any checkpoints it results in the exception You observed. That is because the partial presence of transaction logs files without any checkpoints is the exceptional situation.
But for cases where transaction logs are not present at all that should not be a case.
Please see an attached isolated test case that start-stop the database, delete all present transaction log files, and restart the database without any problems.

```
package org.neo4j;

import org.junit.Rule;
import org.junit.Test;

import java.io.File;

import org.neo4j.graphdb.GraphDatabaseService;
import org.neo4j.graphdb.Transaction;
import org.neo4j.helpers.collection.Iterables;
import org.neo4j.io.fs.FileUtils;
import org.neo4j.kernel.impl.transaction.log.files.LogFiles;
import org.neo4j.kernel.internal.GraphDatabaseAPI;
import org.neo4j.test.TestEnterpriseGraphDatabaseFactory;
import org.neo4j.test.rule.TestDirectory;

import static org.apache.commons.lang3.StringUtils.repeat;
import static org.junit.Assert.assertEquals;
import static org.neo4j.graphdb.Label.label;

public class StartDatabaseWithoutTxLogs
{
    @Rule
    public final TestDirectory testDirectory = TestDirectory.testDirectory();

    @Test
    public void recoveryNoTxLogs()
    {
        File storeDir = testDirectory.graphDbDir();

        GraphDatabaseService database = new TestEnterpriseGraphDatabaseFactory().newEmbeddedDatabase( storeDir );
        try ( Transaction transaction = database.beginTx() )
        {
            database.createNode( label( "test1" ) );
            database.createNode( label( "test2" ) );
            database.createNode( label( "test3" ) );
            transaction.success();
        }

        LogFiles logFiles = ((GraphDatabaseAPI) database).getDependencyResolver().resolveDependency( LogFiles.class );
        File[] txLogFiles = logFiles.logFiles();

        listFilesWithMessage( storeDir, "List files while database is running." );
        database.shutdown();

        for ( File txLogFile : txLogFiles )
        {
            FileUtils.deleteFile( txLogFile );
        }

        listFilesWithMessage( storeDir, "List files before database restart." );
        GraphDatabaseService restartedDatabase = new TestEnterpriseGraphDatabaseFactory().newEmbeddedDatabase( storeDir );
        try ( Transaction transaction = restartedDatabase.beginTx() )
        {
            assertEquals( 3, Iterables.count( restartedDatabase.getAllNodes() ) );
        }
        restartedDatabase.shutdown();

        listFilesWithMessage( storeDir, "List files after restarted database stop." );
    }

    private void listFilesWithMessage( File storeDir, String s )
    {
        System.out.println();
        System.out.println( repeat( "*", 60 ) );
        System.out.println( s );
        listFiles( storeDir );
    }

    private static void listFiles( File directory )
    {
        System.out.println( "========================List files===========================" );
        System.out.println( "Directory: " + directory.getAbsolutePath() );
        File[] files = directory.listFiles();
        if ( files != null )
        {
            for ( File file : files )
            {
                System.out.println( file.getAbsolutePath() );
            }
        }
    }
}
```

Out of curiosity, what was the reason to remove transaction logs manually in the first place?

Thx in advance for Your answers.

**Comment by tutolmin** [comment]
I am experiencing the same issue. After DB start new transaction log was created but it is only 16 bytes in size. Neo4j version 3.5.3

I created a RAM disk in order to put transaction logs there and mounted it to /ramdisk mount point.
Then changed dbms.directories.tx_log=/ramdisk in config and tried to start the DB.

/ramdisk/neostore.transaction.db.9 file was created and there was a log message saying: Caused by: org.neo4j.kernel.impl.store.UnderlyingStorageException: No check point found in any log file from version 9 to 9

**Comment by lazyonemax** [comment]
I have this issue to,Neo4j process be killed manual & delete all transaction logs ,so can not startup now 
Version:3.5.7
Error Message
2019-08-08 02:24:28.782+0000 WARN [o.n.k.NeoStoreDataSource] Exception occurred while starting the datasource. Attempting to close things down. Component 'org.neo4j.kernel.recovery.Recovery@795fd838' failed to initialize. Please see the attached cause exception "No check point found in any log file from version 107 to 107". org.neo4j.kernel.lifecycle.LifecycleException: Component 'org.neo4j.kernel.recovery.Recovery@795fd838' failed to initialize. Please see the attached cause exception "No check point found in any log file from version 107 to 107". at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.init(LifeSupport.java:434) at org.neo4j.kernel.lifecycle.LifeSupport.init(LifeSupport.java:66) at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:102) at org.neo4j.kernel.NeoStoreDataSource.start(NeoStoreDataSource.java:444) at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452) at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:111) at org.neo4j.kernel.impl.transaction.state.DataSourceManager.start(DataSourceManager.java:116) at org.neo4j.kernel.lifecycle.LifeSupport$LifecycleInstance.start(LifeSupport.java:452) at org.neo4j.kernel.lifecycle.LifeSupport.start(LifeSupport.java:111) at org.neo4j.graphdb.facade.GraphDatabaseFacadeFactory.initFacade(GraphDatabaseFacadeFactory.java:225) at com.neo4j.commercial.edition.CommercialGraphDatabase.<init>

**Comment by lazyonemax** [comment]
> Hi @tricepat, thx for Your report.
> 
> I have tried to reproduce reported behaviour over multiple version (different 3.4 and 3.5 versions) and so far I do not see a way that will prevent database to start in the case when transaction logs are not there.
> 
> Also, please note that in the provided stacktrace You can see the message "No check point found in any log file from version 14 to 14". That means that tx log file with index 14 was present during the database start.
> Since that file did not contain any checkpoints it results in the exception You observed. That is because the partial presence of transaction logs files without any checkpoints is the exceptional situation.
> But for cases where transaction logs are not present at all that should not be a case.
> Please see an attached isolated test case that start-stop the database, delete all present transaction log files, and restart the database without any problems.
> 
> ```
> package org.neo4j;
> 
> import org.junit.Rule;
> import org.junit.Test;
> 
> import java.io.File;
> 
> import org.neo4j.graphdb.GraphDatabaseService;
> import org.neo4j.graphdb.Transaction;
> import org.neo4j.helpers.collection.Iterables;
> import org.neo4j.io.fs.FileUtils;
> import org.neo4j.kernel.impl.transaction.log.files.LogFiles;
> import org.neo4j.kernel.internal.GraphDatabaseAPI;
> import org.neo4j.test.TestEnterpriseGraphDatabaseFactory;
> import org.neo4j.test.rule.TestDirectory;
> 
> import static org.apache.commons.lang3.StringUtils.repeat;
> import static org.junit.Assert.assertEquals;
> import static org.neo4j.graphdb.Label.label;
> 
> public class StartDatabaseWithoutTxLogs
> {
>     @Rule
>     public final TestDirectory testDirectory = TestDirectory.testDirectory();
> 
>     @Test
>     public void recoveryNoTxLogs()
>     {
>         File storeDir = testDirectory.graphDbDir();
> 
>         GraphDatabaseService database = new TestEnterpriseGraphDatabaseFactory().newEmbeddedDatabase( storeDir );
>         try ( Transaction transaction = database.beginTx() )
>         {
>             database.createNode( label( "test1" ) );
>             database.createNode( label( "test2" ) );
>             database.createNode( label( "test3" ) );
>             transaction.success();
>         }
> 
>         LogFiles logFiles = ((GraphDatabaseAPI) database).getDependencyResolver().resolveDependency( LogFiles.class );
>         File[] txLogFiles = logFiles.logFiles();
> 
>         listFilesWithMessage( storeDir, "List files while database is running." );
>         database.shutdown();
> 
>         for ( File txLogFile : txLogFiles )
>         {
>             FileUtils.deleteFile( txLogFile );
>         }
> 
>         listFilesWithMessage( storeDir, "List files before database restart." );
>         GraphDatabaseService restartedDatabase = new TestEnterpriseGraphDatabaseFactory().newEmbeddedDatabase( storeDir );
>         try ( Transaction transaction = restartedDatabase.beginTx() )
>         {
>             assertEquals( 3, Iterables.count( restartedDatabase.getAllNodes() ) );
>         }
>         restartedDatabase.shutdown();
> 
>         listFilesWithMessage( storeDir, "List files after restarted database stop." );
>     }
> 
>     private void listFilesWithMessage( File storeDir, String s )
>     {
>         System.out.println();
>         System.out.println( repeat( "*", 60 ) );
>         System.out.println( s );
>         listFiles( storeDir );
>     }
> 
>     private static void listFiles( File directory )
>     {
>         System.out.println( "========================List files===========================" );
>         System.out.println( "Directory: " + directory.getAbsolutePath() );
>         File[] files = directory.listFiles();
>         if ( files != null )
>         {
>             for ( File file : files )
>             {
>                 System.out.println( file.getAbsolutePath() );
>             }
>         }
>     }
> }
> ```
> 
> Out of curiosity, what was the reason to remove transaction logs manually in the first place?
> 
> Thx in advance for Your answers.

sometimes neo4j cannot stop by command ./neo4j stop , will kill process manual (kill -9 pid) , restart load transaction log slowly .......

**Comment by tinwelint** [comment]
This will be fixed in 4.0, but until then you'll have to delete the transaction log, start and shutdown cleanly once to get on track again.
